ECS_MIGRATION.md

```diff
@@ -0,0 +1,93 @@
+
+# Branching
+
+I recommend we have a long-lived `ecs-migration` branch that is spawned **DIRECTLY FROM MASTER**. 
+
+For the moment, I think a single branch will suffice. If we need to manage dev, test, and prod concurrently in both clouds, we can refine this guidance, but it will start to get **much** more complicated at that point, so it will be preferable if we can try to minimize our branches.
+
+# Terraform
+
+## Potential issues
+
+There are a couple of items to worry about:
+- We **do not** want to accidentally overwrite the state of another backend
+- We will be adding new resources, but there will be slight deviations between our current account and ECS, because ECS will not have preexisting resources. For this reason, terraform scripts might drift between `master` and `ecs-migration`.
+
+## Recommended strategy - backends and configs
+
+I recommend we add a backend and a config for the ECS environment as though it was just another environment.
+
+For example:
+```
+/config
+  backend-dev.conf
+  backend-ecs-dev.conf
+  dev.tfvars
+  ecs-dev.tfvars
+```
+
+I recommend `ecs-dev` over `dev-ecs` simply because it would be easy to accidentally type `dev` instead of `dev-ecs` and that could cause problems, whereas a `ecs-dev` wouldn't have that sort of typo problem.
+
+## Recommended strategy - resources
+
+### Greenfield resources
+
+Generally, we will be adding new resources (as ECS will need the resources, but our current account has preexisting manually created resources), so it should be possible to do something like this:
+```
+resource "aws_s3_bucket" "webportal_bucket" {
+  bucket = "webportal-bucket"
+}
+```
+
+This would be safe to run in both clouds, even though it will create a useless bucket in our current account.
+The only issue here will be if we **need to maintain the same name**. I am currently recommending we use new names to avoid complexity. If we MUST use the same name, we'll either have to import the resource into terraform or pass in name as a variable, in which case our current account will get a garbage name, and the new account will get the correct one.
+
+### References and drift
+
+The next bit of complexity is references. At the moment, we will have resources in our account that take variables, and all we should need to do is update them to use a reference instead.
+
+For example
+
+`master` will have:
+```
+resource "aws_cognito_identity_provider" "SamlDotActiveDirectoryIdp" {
+  user_pool_id  = var.user_pool_id
+  provider_name = "${var.environment}-dot-ad"
+  provider_type = "SAML"
+}
+
+# New pool that is not used in our current account, but mirrors the settings of var.user_pool_id pool
+# For testing, we can update the user_pool_id to be the ID of this resource, but DO NOT use a ref above
+resource "aws_cognito_user_pool" "new_pool" {
```

Comment:
What do you think about these existing resources getting imported into Terraform in our public account? If we used this strategy we wouldn't need a long lived ECS branch and risk divergence between master and `ecs-migration`.

Perhaps I'm missing your point...

Reply:
Just being risk averse. We could try that, but I'm a bit worried we'd botch one or more settings. Perhaps create it as a one-off so we can compare with the existing one, and then if they do appear identical, try an import?

Reply:
Yes that works. Lets try it out for the initial story you are picking up.